src/databricks/labs/ucx/azure/azure_credentials.py

```diff
@@ -0,0 +1,283 @@
+        # print action plan to console for customer to review.
+        for sp in sp_list_with_secret:
+            print(
+                f"Service Principal name: {sp.service_principal.principal}, "
+                f"application_id: {sp.service_principal.client_id}, "
+                f"privilege {sp.service_principal.privilege} "
+                f"on location {sp.service_principal.prefix}"
+            )
+
+    def _generate_migration_list(self):
+        """
+        Create the list of SP that need to be migrated, output an action plan as a csv file for users to confirm
+        """
+        # load sp list from azure_storage_account_info.csv
+        sp_list = self._azure_resource_permissions.load()
+        # list existed storage credentials
+        sc_set = self._list_storage_credentials()
+        # check if the sp is already used in UC storage credential
+        filtered_sp_list = [sp for sp in sp_list if sp.client_id not in sc_set]
+        # fetch sp client_secret if any
+        sp_list_with_secret = self._fetch_client_secret(filtered_sp_list)
+        self._final_sp_list = sp_list_with_secret
+        # output the action plan for customer to confirm
+        self._print_action_plan(sp_list_with_secret)
+
+    def _create_storage_credential(self, sp_migration: ServicePrincipalMigrationInfo):
+        # prepare the storage credential properties
+        name = sp_migration.service_principal.principal
+        azure_service_principal = AzureServicePrincipal(
+            directory_id=sp_migration.service_principal.directory_id,
+            application_id=sp_migration.service_principal.client_id,
+            client_secret=sp_migration.client_secret,
+        )
+        comment = f"Created by UCX during migration to UC using Azure Service Principal: {sp_migration.service_principal.principal}"
+        read_only = False
+        if sp_migration.service_principal.privilege == Privilege.READ_FILES.value:
+            read_only = True
+        # create the storage credential
+        storage_credential = self._ws.storage_credentials.create(
+            name=name, azure_service_principal=azure_service_principal, comment=comment, read_only=read_only
+        )
+
+        validation_result = self._validate_storage_credential(
+            storage_credential, sp_migration.service_principal.prefix, read_only
+        )
+        return validation_result
+
+    def _validate_storage_credential(
+        self, storage_credential, location: str, read_only: bool
+    ) -> StorageCredentialValidationResult:
+        # storage_credential validation creates a temp UC external location, which cannot overlap with
+        # existing UC external locations. So add a sub folder to the validation location just in case
+        try:
+            validation = self._ws.storage_credentials.validate(
+                storage_credential_name=storage_credential.name, url=location, read_only=read_only
```

Comment:
Can you use positional arguments instead of named?

Reply:
Not for this one. `_ws.storage_credentials.validate` requires keyword arguments.